brigade/plugins/tasks/networking/netmiko_tasks.py

```diff
@@ -0,0 +1,180 @@
+            net_connect = brigade_host.get_connection("netmiko", **connect_kwargs)
+            kwargs["netmiko_conn"] = net_connect
+        return func(task, **kwargs)
+    return wrapper
+
+
+def netmiko_args(host, ip=None, netmiko_host=None, username=None, password=None, device_type=None,
+                 netmiko_dict=None):
+    """Process arguments coming from different sources."""
+
+    # Both Netmiko and brigade have a host argument
+    brigade_host = host
+    parameters = {
+        "username": username or brigade_host.username,
+        "password": password or brigade_host.password,
+    }
+    if netmiko_host is None and ip is None:
+        parameters["host"] = brigade_host.host
+    elif ip is not None:
+        parameters["ip"] = ip
+    elif netmiko_host is not None:
+        parameters["host"] = netmiko_host
+
+    if netmiko_dict is not None:
+        parameters.update(netmiko_dict)
+    else:
+        netmiko_dict = {}
+    parameters['port'] = netmiko_dict.get('port', brigade_host.ssh_port)
+
+    if device_type is None:
+        if netmiko_dict.get("device_type") is None:
+            device_type = brigade_host.nos
+        else:
+            device_type = netmiko_dict.get("device_type")
+
+    # Convert to netmiko device_type format (if napalm format is used)
+    parameters['device_type'] = napalm_to_netmiko_map.get(device_type, device_type)
+    return parameters
+
+
+@netmiko_get_connection
+def netmiko_run(task, method, ip=None, host=None, username=None, password=None,
+                device_type=None, netmiko_dict=None, cmd_args=None, cmd_kwargs=None,
+                netmiko_conn=None):
+    """
+    Execute any Netmiko method from connection class (BaseConnection class and children).
+
+    Arguments:
+        method(str): Netmiko method to use
+        ip (string, optional): defaults to ``brigade_ip``
+        host (string, optional): defaults to None
+        username (string, optional): defaults to ``brigade_username``
+        password (string, optional): defaults to ``brigade_password``
+        device_type (string, optional): Netmiko device_type to use, defaults to ``nos`` (mapped \
+        through napalm_to_netmiko_map)
+        netmiko_dict (dict, optional): Additional arguments to pass to Netmiko ConnectHandler, \
+        defaults to None
+
+    Returns:
+        :obj:`brigade.core.task.Result`:
+          * result (``dict``): dictionary with the result of the getter
+    """
+    # Netmiko connection automatically created by 'netmiko_get_connection' decorator (if needed)
+    netmiko_method = getattr(netmiko_conn, method)
+    if cmd_args is None:
+        cmd_args = ()
+    if cmd_kwargs is None:
+        cmd_kwargs = {}
+    result = netmiko_method(*cmd_args, **cmd_kwargs)
+    return Result(host=task.host, result=result)
+
+
+@netmiko_get_connection
+def netmiko_send_command(task, ip=None, host=None, username=None, password=None,
```

Comment:
What's the use case for duplicating all the connection parameter on all the tasks? Why can't the user call `netmiko_connection` if some customization is needed and then just call `netmiko_send_command`? I think we should handle connection only in the connection task to avoid duplicating parameters in many tasks (that could lead to inconsistencies if we don't update all the tasks when something changes) and to avoid bloating the documentation with lots of parameters.

Reply:
I think our philosophy here basically should be to have lots of small and simple tasks rather than few and very complex. I'd prefer to have smaller and simpler workflows than end up writing this monster again:

https://github.com/napalm-automation/napalm-ansible/blob/develop/napalm_ansible/napalm_install_config.py

IMHO, because we are dealing with python code directly we are in a good position to make things simpler to write, maintain and understand.

Reply:
The use was to have the task automatically setup the connection if it didn't exist...so end-users didn't have the extra step of the connection setup call.

I would say this should be a prerogative of the task...i.e. Brigade shouldn't care.

That all being said...I will change it as the argument management is getting ugly.